### PR TITLE
Colorizable smokers

### LINKED ISSUE #656: [Suggestion] Colourizable smokers
How about adding an option to the smokers that would enable changing the smoke color? Like, right-clicking with a dye, or having a dye somewhere in the crafting recipe? Not all smoke is black :)

### PR DESCRIPTION
**Pull request by liach**
Closes #656
Smokers now can be colored by using any ore dictionary dye on the smoker
with main hand while not sneaking. It consumes one dye. Adds support for
comparator output for smoker color.
There are some minor fixes needed for inventories, too.

### CODE REVIEW

**Comment by liach** [comment]
Will get a screenshot soon.

**Comment by liach** [comment]
Smokers and redstone output

**Comment by Generalcamo** [comment]
I'm not entirely sure why a smoker needs to output comparator redstone signals based on color. But I suppose you could say the same for Jukeboxes and the like.

File: src/main/java/mods/railcraft/common/util/inventory/InvTools.java
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/util/inventory/InvTools.java]: This is basically an incomplete version of depleteItem(ItemStack).

Comment (liach) [reply src/main/java/mods/railcraft/common/util/inventory/InvTools.java]: For deplete one, does it mutate the stack itself? I don't understand it very well so I rewrote a size decrease function that always changes the stack in the argument.

Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/util/inventory/InvTools.java]: Mutating the stack is a bad idea generally. I've gotten tons of flack for doing that in the past from other modders who use object identity to detect changes to stacks.

Comment (liach) [reply src/main/java/mods/railcraft/common/util/inventory/InvTools.java]: Most of the uses of `ItemStack#stackSize` field are direct mutations, so it is not that bad an idea. In this case, this stack is in player's hand, consuming one after dying the smoker.

Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/util/inventory/InvTools.java]: But I guess it might have its uses.

Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/util/inventory/InvTools.java]: However in the case of the Smoker, it should definitely use depleteItem() for the container support.

**Comment by liach** [comment]
@CovertJaguar Build was killed by missing `BlockChargeBattery.java`, not caused by pr contents.

File: src/main/java/mods/railcraft/common/blocks/machine/equipment/TileSmoker.java
Comment (Generalcamo) [inline src/main/java/mods/railcraft/common/blocks/machine/equipment/TileSmoker.java]: Use InvTools.isEmpty(heldItem)

File: src/main/java/mods/railcraft/common/blocks/machine/equipment/TileSmoker.java
Comment (Generalcamo) [inline src/main/java/mods/railcraft/common/blocks/machine/equipment/TileSmoker.java]: Seems like a large amount of spacing added here.

**Review by Generalcamo — DISMISSED** [review]

**Review by Generalcamo — APPROVED** [review]
Wait for CJ's approval, but I see no issues.

**Review by CovertJaguar — APPROVED** [review]

**Pull request merged**
